### DISCUSSION TITLE
Installing MOOSE on ARM64 architecture

### DISCUSSION THREAD

**Discussion by shikhar413** [idaholab/moose#17416] [Q&A Getting Started] [answered]
Hello MOOSE team,

I am a new MOOSE user and I am trying to build MOOSE on an ARM64 machine. I've had to rely on building most dependencies from source since many conda packages are built only on x86_64 machines. When compiling PETSC and libMesh by running `./scripts/update_and_rebuild_petsc.sh` and `./scripts/update_and_rebuild_libmesh.sh`, however, I run into the following messages:

PETSC Compilation Error:
![image](https://user-images.githubusercontent.com/14334316/112175296-87edd780-8bcd-11eb-81a0-9e59783bb296.png)

libMesh Compilation Error:
![image](https://user-images.githubusercontent.com/14334316/112176188-43af0700-8bce-11eb-8e6d-44fa70bfe05b.png)

These errors occur while running `make` after `configure` has finished being called. I am running on ARM64 Ubuntu Server 20.04 and have installed MOOSE dependencies according to the instructions on https://mooseframework.inl.gov/getting_started/installation/manual_installation_gcc.html. I don't know how I can modify the PETSC and LibMesh install scripts to get around these issues, any guidance would be very much appreciated.

**Comment by aeslaughter** [comment]
@dschwen You got MOOSE running on a Raspberry PI, correct? Do you have any insights?

**Comment by shikhar413** [comment, marked as answer]
The folks at Argonne MCS were able to lead me in the right direction wrt the PETSc installation. For some reason the Fortran compiler was unable to locate the shared libraries for mumps and manually setting the PETSc flag in `scripts/update_and_rebuild_petsc.sh` as `FFLAGS="-fPIC -Wall -ffree-line-length-0 -Wno-unused-dummy-argument"` fixed the PETSc compilation issue.

Adding the full commands to build MOOSE on ARM Ubuntu 20.04 for anyone else who might run into similar issues, happy to submit a PR to add these instructions to some documentation in the repo if there is interest as well. Might be good to also start testing builds on ARM architectures, especially to support installation on the new M1 Macs.

```bash
#!/bin/sh

# Install gcc, python3
sudo apt install build-essential libssl-dev libffi-dev python3-dev manpages-dev

# Install MPICH
sudo apt install mpich

# Install pip
sudo apt install python3-pip

# Install MOOSE dependencies
pip3 install coverage
pip3 install reportlab
pip3 install mako
pip3 install numpy
pip3 install scipy
pip3 install scikit-learn
sudo apt install hdf5-tools
sudo apt install hdf5-helpers
sudo apt install python3-h5py
pip3 install scikit-image
pip3 install requests
sudo apt install python3-vtk7
sudo apt install python3-vtkplotter
pip3 install pyyaml
pip3 install matplotlib
pip3 install lxml
pip3 install pyflakes
pip3 install pandas
pip3 install mock
sudo apt install python3-yaml
sudo apt install python3-pyqt5
sudo apt install swig
pip3 install --no-cache-dir pybtex livereload daemonlite pylint lxml pylatexenc anytree

# Add locally installed binaries to path
echo "export PATH=$HOME/.local/bin:$PATH" >> ~/.bashrc

# Set environment variables for MOOSE installation
echo "export CC=mpicc" >> ~/.bashrc
echo "export CXX=mpicxx" >> ~/.bashrc
echo "export FC=mpif90" >> ~/.bashrc
echo "export F90=mpif90" >> ~/.bashrc

# Point default python environment to python3 (necessary for running MOOSE tests)
sudo apt-get install python-is-python3

# Install PETSC dependencies
sudo apt install bison valgrind flex

# Clone MOOSE repo
moose_dir=$HOME/projects
mkdir -p $moose_dir
cd $moose_dir
git clone https://github.com/idaholab/moose.git
cd moose
git checkout master

# Manually add flag for Fortran flags to PETSc installation script
sed -i 's/hypre=1/hypre=1 --FFLAGS="-fPIC -Wall -ffree-line-length-0 -Wno-unused-dummy-argument"/g' scripts/update_and_rebuild_petsc.sh

# Compile PETSc
unset PETSC_DIR PETSC_ARCH
./scripts/update_and_rebuild_petsc.sh
cd petsc
make PETSC_DIR=$moose_dir/moose/scripts/../petsc PETSC_ARCH=arch-moose check

# Compile libMesh
cd $moose_dir/moose
./scripts/update_and_rebuild_libmesh.sh

# Compile and test MOOSE
cd $moose_dir/moose/test
make -j 4
./run_tests -j 4
```
